### PR TITLE
Generate SBOM for istio repository

### PR DESCRIPTION
**Pull request by kfaseela**
Attempt to generate SBOM for istio repository in  SPDX format.
This PR integrates the Bill of Materials code Into the istio release process. The current PR generates the first SBOM file that only describes the istio source code. The complete plan is to generate one SBOM for the source code and one for the release artifacts. 

kbs/bom utility is in the process of publishing a new release soon, which will have enhanced functionalities that would help us generate BOM for docker image-archives, and other tarballs in the release. Waiting for that to add the remaining BOM generation portion for release artifacts. 

RFC : [SBOM for Istio](https://docs.google.com/document/d/10ITZUUf-aGbeiBKfr-EqrQTPfPpVQmUT/)

Signed-off-by: Faseela K <faseela.k@est.tech>

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Nice! Looks close. Two comments

1. Need to decide between this tool and kubernetes-sigs/bom.
2. Whichever one we pick we should add it to the build-tools container in Istio/tools repo, not run under docker build

I reached out to k8s team to learn more about `bom`

**Comment by kfaseela** [comment]
> Nice! Looks close. Two comments
> 
> 1. Need to decide between this tool and kubernetes-sigs/bom.
> 2. Whichever one we pick we should add it to the build-tools container in Istio/tools repo, not run under docker build
> 
> I reached out to k8s team to learn more about `bom`

Thanks @howardjohn , I will also try to run both the utilities locally and see if there is a difference in the output, so that we can compare. will wait for your reply before pushing a revised PR

**Comment by howardjohn** [comment]
Did a bit of digging into the `bom` tool and talking with K8s folks (but did NOT run the tools), and so far I prefer it.

One thing I noticed is we are just generating the istio/istio repo. But we have many things we publish.. do we need something like https://github.com/kubernetes-sigs/bom/blob/main/docs/create-a-bill-of-materials.md#generating-the-sbom

**Comment by kfaseela** [comment]
> Did a bit of digging into the `bom` tool and talking with K8s folks (but did NOT run the tools), and so far I prefer it.
> 
> One thing I noticed is we are just generating the istio/istio repo. But we have many things we publish.. do we need something like https://github.com/kubernetes-sigs/bom/blob/main/docs/create-a-bill-of-materials.md#generating-the-sbom

https://sbom.k8s.io/v1.23.0/release
https://sbom.k8s.io/v1.23.0/source

we would need something similar. one for release and one for source. I am looking into more details and in touch with k8s-sig/bom team

File: pkg/build/sbom.go
Comment (kfaseela) [inline pkg/build/sbom.go:31]: this needs to change

File: pkg/build/sbom.go
Comment (puerco) [inline pkg/build/sbom.go:34]: OK, we don't (yet) have a way to link these two from the command line (you can do it from the library). Hopefully, we can tackle that in an upcoming release of `bom`.

File: pkg/build/sbom.go
Comment (puerco) [inline pkg/build/sbom.go:39]: I thought the image archives were going to be included here?

Comment (kfaseela) [reply pkg/build/sbom.go:39]: will include --image-archive once v0.2.0 of bom is released. Currently our build-tools is on 0.1.0. Until then docker image tar balls are also scanned as files.

Comment (puerco) [reply pkg/build/sbom.go:39]: Cool, we are [one PR away](https://github.com/kubernetes-sigs/bom/pull/31) from 0.2, it should be out shortly!

Comment (kfaseela) [reply pkg/build/sbom.go:55]: @puerco : upgraded to 0.2.0 and using --image-archive now!

**Comment by puerco** [comment]
It seems the internal structure of the layers changed, we made some changes to handle it, it should run once this merges:
https://github.com/kubernetes-sigs/bom/pull/37

**Comment by kfaseela** [comment]
> It seems the internal structure of the layers changed, we made some changes to handle it, it should run once this merges: [kubernetes-sigs/bom#37](https://github.com/kubernetes-sigs/bom/pull/37)

Thanks @puerco . will wait for 0.2.1 tag with the fix.

**Comment by puerco** [comment]
Retrying the test in case istio/tools#1846 needed some time to soak
/test dry-run_release-builder

**Comment by kfaseela** [comment]
@ericvn @puerco  any idea why this PR is not picking up latest bom version which we merged as part of https://github.com/istio/tools/pull/1846 ?

**Comment by howardjohn** [comment]
needs to go through https://github.com/istio/common-files/pull/524 first, I approved it

**Comment by kfaseela** [comment]
> [istio/tools#1846](https://github.com/istio/tools/pull/1846)

Thank you!

**Comment by puerco** [comment]
Thanks @howardjohn, I can confirm I see the version we need in the `gcr.io/istio-testing/build-tools:master-2022-01-30T03-45-42` image :+1:

**Comment by kfaseela** [comment]
> Did a bit of digging into the `bom` tool and talking with K8s folks (but did NOT run the tools), and so far I prefer it.
> 
> One thing I noticed is we are just generating the istio/istio repo. But we have many things we publish.. do we need something like https://github.com/kubernetes-sigs/bom/blob/main/docs/create-a-bill-of-materials.md#generating-the-sbom

Done.

**Comment by puerco** [comment]
Great work @kfaseela ! Looks like a great SBOM is coming in! 👍

/lgtm

**Review by puerco — APPROVED** [review]

File: pkg/build/sbom.go
Comment (ericvn) [inline pkg/build/sbom.go:34]: Is this the correct URL? It does make sense to have the published URL, but I see that as something like `https://gcsweb.istio.io/gcs/istio-release/releases/1.12.2/`, but that doesn't seem to map to the storage.googleapis.com.

File: pkg/build/sbom.go
Comment (ericvn) [inline pkg/build/sbom.go:37]: Same URL question.

File: pkg/build/sbom.go
Comment (kfaseela) [reply pkg/build/sbom.go:34]: Given URL works for manifest : https://storage.googleapis.com/istio-release/releases/1.12.2/manifest.yaml and since spdx files are published at the same location as manifest.yaml, I am assuming this would be the right URL once spdx files are published.

**Review by ericvn — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**
